api/handlers/metadata.py

```diff
@@ -0,0 +1,91 @@
+import database.database
+from bson.errors import InvalidId
+from bson.objectid import ObjectId
+from fastapi import APIRouter, Body, Response, Depends
+
+router = APIRouter()
+
+@router.get('/api/datasources/{datasource_id}/meta', status_code = 200)
+def get_all_meta(datasource_id, response: Response, db: object = Depends(database.database.db)):
+    
+    # TODO: Should we validate datasource_id?
+    
+    # Return all metadata for this datasource, could be an empty
+    # list
+    metadata = db.metadata.find({'datasource_id': datasource_id})
+    result = []
+    for datum in metadata:
+        datum['_id'] = str(datum['_id'])
+        result.append(datum)
+    return result
+
+@router.get('/api/datasources/{datasource_id}/{filepath}/meta', status_code = 200)
+def get_meta(datasource_id, filepath, response: Response, db: object = Depends(database.database.db)):
+    metadata = db.metadata.find_one({'datasource_id': datasource_id, 'filepath': filepath})
+    if not metadata:
+        response.status_code = 404 
+        return "Not Found"
+    metadata['_id'] = str(metadata['_id'])
+    return metadata
+
+@router.post('/api/datasources/{datasource_id}/{filepath}/meta', status_code = 201)
+def create_meta(datasource_id, filepath, response: Response, db: object = Depends(database.database.db), metadata = Body(...)):
+
+      # Check datasource id is valid
+      try:
+        accountName, containerName = datasource_id.split("_")
+        datasource = db.datasources.find_one({"accountName" : accountName, "containerName" : containerName})
+        if not datasource:
+            response.status_code = 404
+            return "Datasource Not Found"
+      except ValueError: # No '_' in datasource_id
+          response.status_code = 400
+          return "Invalid Datasource Id"
+
+      # Check metadata doesn't already exist
+      if db.metadata.find_one({'datasource_id': datasource_id, 'filepath': filepath}):
+          response.status_code = 400
+          return {"error":"record already exists"}
+
+      # Create the first metadata record
+      initial_version = {
+          'version_number': 0,
+          'datasource_id': datasource_id,
+          'filepath': filepath,
+          'metadata': metadata
+      }
+      result = db.metadata.insert_one(initial_version)
+      db.versions.insert_one(initial_version)
+      return "Success"
+
+def get_latest_version(db, datasource_id, filepath):
+    cursor = db.versions.find({'datasource_id': datasource_id, 'filepath': filepath}).sort('version', -1).limit(1)
```

Comment:
Why on this method not just use the db.metadata?

Reply:
Syntax Error. Can you explain a bit more.

Reply:
As broad as it is long.